code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
+        create.removeAll(update);
+        create.removeAll(createIgnore);
+        HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+        delete.removeAll(deleteIgnore);
+
+        Condition allMembers = DSL.falseCondition();
+        for (Member member: delete){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+        }
+        List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+        for (ProjectMember member: toDelete){
+            objectProcessManager.scheduleStandardProcess(StandardProcess.DEACTIVATE, member, null);
+            objectProcessManager.scheduleStandardProcess(StandardProcess.REMOVE, member, null);
+        }
+        List<ProjectMember> createdMembers = new ArrayList<>();
+        for (Member member: create){
+            createdMembers.add(createProjectMember(project, member));
+        }
+        for (Member member: update){
+            createdMembers.add(updateMember(project, member));
+        }
+        //get all members that you need to update
+        //create all members that did not exist
+        return getProjectMembers(project.getId());
     }
 
-    public GenericResourceDao getGenericResourceDao() {
-        return resourceDao;
+    private ProjectMember updateMember(Account project, Member member) {
+        ProjectMemberRecord gotMember = create()
+                .selectFrom(PROJECT_MEMBER).where(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId()))
+                .and(PROJECT_MEMBER.REMOVED.isNull())).fetchOne();
+        gotMember.setRole(member.getRole());
+        objectManager.persist(gotMember);
+        return gotMember;
     }
 
-    @Inject
-    public void setGenericResourceDao(GenericResourceDao resourceDao) {
-        this.resourceDao = resourceDao;
+    ProjectMember createProjectMember(Account project, Member member) {
+        return creatProjectMember(project, new ExternalId(member.getExternalId(), member.getExternalIdType()), member.getRole());
+    }
+
+    public ProjectMember creatProjectMember(Account project, ExternalId externalId, String role) {
+        Map<Object, Object> properties = new HashMap<>();
+        properties.put(PROJECT_MEMBER.PROJECT_ID, project.getId());
+        properties.put(PROJECT_MEMBER.ACCOUNT_ID, project.getId());
+        properties.put(PROJECT_MEMBER.NAME, project.getName());
+        properties.put(PROJECT_MEMBER.EXTERNAL_ID, externalId.getId());
+        properties.put(PROJECT_MEMBER.EXTERNAL_ID_TYPE, externalId.getType());
+        properties.put(PROJECT_MEMBER.ROLE, role);
```

Comment:
This is so that i don't need to wait for them to be active to get them as i return them after creation within the same api request. And they will not be active otherwise. As a result the user will not see them when updating a project.